### DISCUSSION TITLE
Small component taking a long time

### DISCUSSION THREAD

**Discussion by selfsimilar** [livewire/livewire#5572] [1. Help] [answered]
I have implemented a simple toggle button in Livewire, of which there are 52 instances on a given page. Without the toggles the page takes < 0.5s to render, but with the toggles it jumps to over 3 seconds. I've been careful to avoid DB calls on the component by eager-loading the data it needs. Telescope shows only 19 queries total (duration 34ms) regardless of whether the Livewire components are included or excluded. I just updated to Laravel 10, but this was happening under 9 as well. I'm running this locally via Valet with full MariaDB and Redis for cache.

The only thing I can think of is that I'm passing the full objects with eager-loaded relationships to the components, so the parameters are pretty large, and are currently public properties. This will bloat the page a bit, but cuts down on DB queries. Could that really explain almost 50ms per component added to page load time?

Code below:
Foo and Bar have a many-to-many relationship.

```php
namespace App\Http\Livewire;

use App\Http\Livewire\Toggle;
use App\Models\Foo;
use App\Models\Studio;

class FooBarToggle extends Toggle
{
    public Foo $foo;
    public Bar $bar;

    public function initializeState()
    {
        $this->is_active = (bool) $this->foo->bars->contains($this->bar);
        parent::initializeState();
    }

    public function toggle()
    {
        parent::toggle();

        $action = $this->is_active ? 'attach' : 'detach';
        $this->foo->bars()->$action($this->bar->id);
    }
}
```

```php
namespace App\Http\Livewire;

use Livewire\Component;

class Toggle extends Component
{
    public bool $is_active = false;
    public string $event = 'toggle';
    public string $label = '';

    public function mount()
    {
            $this->initializeState();
    }

    public function initializeState()
    {
    }

    public function toggle()
    {
        $this->is_active = ! $this->is_active;
        $this->emitUp($this->event);
    }

    public function render()
    {
        return view('livewire.toggle');
    }
```

```blade
<div>
    <div wire:click="toggle" class="cursor-pointer flex items-center">
        <div class="inline-block rounded-2xl p-0.5 w-14 h-7 @if ($is_active) bg-green-500 @else bg-gray-400 @endif transition">
            <div class="bg-white rounded-xl w-6 h-6 transition @if ($is_active) ml-7 @endif"></div>
        </div>
        <div class="inline-block font-bold ml-2">{{ $label }}</div>
     </div>
</div>
```

```blade
<div>
    @foreach ($bars as $bar)
        @livewire ('foo-bar-toggle', ['foo' => $foo, 'bar' => $bar], key("{$foo->id}-{$bar->id}"))
    @endforeach
</div>
```

**Comment by joshhanley** [comment]
@selfsimilar hard to say whats going on as I can't see how you are getting `$bars` in your parent component and it would seem like that is the important piece.

**Comment by selfsimilar** [comment, marked as answer]
I refactored and page load is now about 600ms. I no longer pass the full objects, just the ids and the initial state.

But for the record, `$this->foo->bars` should have been eager loaded and the relationship was defined in `App\Models\Foo` with:

```php
public function bars()
{
    return $this->belongsToMany(Bar::class);
}
```

The refactored code looks like:

```php
namespace App\Http\Livewire;

use App\Http\Livewire\Toggle;
use App\Models\Foo;

class FooBarToggle extends Toggle
{
    public int $foo_id;
    public int $bar_id;

    public function toggle()
    {
        parent::toggle();

        $action = $this->is_active ? 'attach' : 'detach';
        $foo = Foo::find($this->foo_id);
        $foo->bars()->$action($this->bar_id);
    }
}
```

I create an array that has the toggle state so I can pass that to the component directly in `Page.blade.php` (below). I was really only passing the whole objects so that the component had the responsibility of looking up state. But because I had to pass the objects as public properties, it was dehydrating some pretty big objects that were eager-loaded with plenty of relationship data, which I think was what was eating up so much time.

Page.blade.php
```blade
<div>
    @foreach ($bars as $bar)
        @livewire ('foo-bar-toggle', [
            'foo_id' => $foo->id,
            'bar_id' => $bar->id,
            'is_active' => in_array($bar->id, $active_bars_in_foo),
        ], key("{$foo->id}-{$bar->id}"))
    @endforeach
</div>
```

**Comment by selfsimilar** [comment]
I was having a similar load issue on a different component that was not as easy to refactor, and I discovered the following technique, if it helps anyone:

```php
namespace App\Http\Livewire;

use App\Models\BigEagerFoo;
use Livewire\Component;

class Foo extends Component
{
    public int $foo_id;

    private Foo $foo;

    public function mount(Foo $foo)
    {
        $this->foo = $foo;
        $this->foo_id = $foo->id;
    }

    public function render()
    {
        $this->hydrate();
        return view('livewire.foo', ['foo' => $this->foo]);
    }

    private function hydrate()
    {
        $this->foo ??= Foo::find($this->foo_id);
    }

}
```

This will allow your page controller to load a bunch of Foo objects and then pass them on fully-loaded to the Livewire component, without the component needing to worry about dehydrating/serializing big objects, especially with eager-loaded relationship data. This is obviously not helpful for highly interactive components, but especially for modal windows that are mostly just displaying static content (like how Jetstream implements its modals in Livewire), it reduces load time significantly.

**Reply by joshhanley** [reply]
@selfsimilar a couple of suggestions, don't call `hydrate()` in the render method as that is a Livewire lifecycle method so Livewire will call that for you on each subsequent request, just not the initial request (but you shouldn't need it anyway). Or rename it to something else if you want to do it that way.

The other thing is to make use of computed properties instead of the private property. So the component would look like this:

```php
namespace App\Http\Livewire;

use App\Models\BigEagerFoo;
use Livewire\Component;

class Foo extends Component
{
    public int $foo_id;

    public function getFooProperty()
    {
        return Foo::find($this->foo_id);
    }

    public function mount(Foo $foo)
    {
        $this->foo_id = $foo->id;
    }

    public function render()
    {
        return view('livewire.foo');
    }
}
```

This way you can access the computed property using `$this->foo` both in the component class and view.

Computed properties are also memoized, so the query will only run the first time the computed property is called (within a single request, it re-runs on each request).

Hope this helps!

**Reply by selfsimilar** [reply]
Thanks for the note about `hydrate()` – I'll avoid using that as a function name.

And while the computed property is good, it's not a great solution by itself to avoid the n+1 problem. For example, if I have a `Foo` dashboard page that shows 50 `Foo` objects at a time, which is the parent view, it makes much more sense load and iterate over the full objects to reduce DB queries. But using a computed property as demonstrated that will drive the DB queries back up again.

You can get the best of both worlds with the following technique. Create and initialize on `mount()` a private variable with a different name than the Foo property, to avoid shadowing. This allows you to use the fully-loaded object on first call, but subsequent calls will make the necessary DB call.

```php
namespace App\Http\Livewire;

use App\Models\Foo;
use Livewire\Component;

class Foo extends Component
{
    public int $foo_id;

    private Foo $_foo;

    public function getFooProperty()
    {
        return $this->_foo ?? Foo::find($this->foo_id);
    }

    public function mount(Foo $foo)
    {
        $this->foo_id = $foo->id;
        $this->_foo = $foo;
    }

    public function render()
    {
        return view('livewire.foo');
    }
}
```

**Reply by joshhanley** [reply]
@selfsimilar oh nice! Yep forgot about that part. I wonder whether you can "preload" a computed property, removing the need for the private property 🤔

**Reply by selfsimilar** [reply]
Looking under the hood a bit, it turns out you're right. You can forego the `private` property by directly setting `$this->computedPropertyCache` in `mount()`. It feels a little funny setting a `protected` property this way, but, it's slightly simpler and it works.

```php
namespace App\Http\Livewire;

use App\Models\Foo;
use Livewire\Component;

class Foo extends Component
{
    public int $foo_id;

    public function getFooProperty()
    {
        return Foo::find($this->foo_id);
    }

    public function mount(Foo $foo)
    {
        $this->foo_id = $foo->id;
        $this->computedPropertyCache['foo'] = $foo;
    }

    public function render()
    {
        return view('livewire.foo');
    }
}
```

**Reply by selfsimilar** [reply]
This might be too 'magic', and I wouldn't want to try to implement this on the current version of Livewire, but I can imagine adding a method to `Livewire\Component` that runs on boot that introspects the object to find all computed properties (e.g. all public methods that match `get[A-Z].*Property` regex), and then tries to set the associated `computedPropertyCache[]` if a variable whose name matches the property is being passed to component. I don't even know how possible this is without further digging, and it really only saves one line per property in the `mount()` function, but at least it protects a user from having to know about the internals of and set the `computedPropertyCache` array.

**Reply by joshhanley** [reply]
@selfsimilar nice! Yeah that might be too much magic, but a helper method to preload the computed property could be handy.
